### DISCUSSION TITLE
Linter 2023 Focus: stability and user experience

### DISCUSSION THREAD

**Discussion by Conaclos** [biomejs/biome#5] [Ideas]
This is an in-progress report outlining the focus for the next few weeks.

### Promote some suggestions to safe fixes

A suggestion or unsafe fix can change the behavior of the program or the intent of the programmer.
In contrast to suggestions, A safe fix does not change the behavior of the program.
A user can automatically apply a safe fixes with confidence.
From a rule maker point of view is thus scary of marking a fix as a safe fix.
Thus, we generally start with suggestions.

Biome has now more rules with suggestions than safe fixes.
This leads some users to directly apply all unsafe fixes.
As with nursery rules which are promoted to another group once stabilized,
we should promote **relevant** suggestions to safe fixes once we are confident about them.

I think that the following rules could benefit of the promotion of (some of) their suggestions to safe fixes.

- [x] [noMultipleSpacesInRegularExpressionLiterals] (delayed because of [#383](https://github.com/biomejs/biome/issues/383))
- [x] [noNegationElse]
- [x] [noUselessLabel]
- [x] [noUselessTypeConstraint]
- [x] [noUnusedLabels]
- [x] [useConst]
- [x] [useEnumInitializers]
- [ ] [useExponentiationOperator]
- [ ] [useLiteralKeys]
- [ ] [useNumericLiterals]
- [ ] [useOptionalChain]
- [ ] [useSingleVarDeclarator] (delayed because of #728)
- [ ] [useTemplate]
- [x] [useWhile]

### Provide more code fixes

- [ ] [noUnusedVariables]

  A `catch` with an unused variable could be shorten to an argument-less `catch`.

- [x] [useIsNan]

  We could provide a suggestion.

### Smarter rules

- [x] [useTemplate]

  The rule provides the following suggestion:

  ```diff
  - a + b + "px"
  + `${a}${b}px`
  ```

  This breaks code where `a` and `b` are numbers ([rome#4713](https://github.com/rome/tools/issues/4713)).
  To fix the issue, we could make the following suggestion:

  ```diff
  - a + b + "px"
  + `${a + b}px`
  ```

  Note that the issue is not present for the following code:

  ```js
  "s" + a + b
  ```

- [x] Remove [noAssignInExpressions] code fix

  The code fix is sometimes unwanted by users [rome#4717](https://github.com/rome/tools/issues/4717)

### Deprecate annoying or confusing rules

- [x] Remove [noConfusingArrow]

  The rule is already turned off in most cases.

- [ ] [useSingleCaseStatement]

  The rule is confusing: it requires block around in switch clause if there is more than 1 statement.
  A lot of users complained about the rule.

  We should deprecate the rule and provide a rule that enforces block in all switch clauses.

### Write guidelines

**Rule promotions.** Guideline to promote a rule (when, ...). 

**Code fixes.** It is sometimes hard to decide whether a code fix is safe or unsafe.
Safe fixes are automatically applied with the fix-all and quickfixes on-save IDE actions.
As a result, it is a bit risky of marking a code fix as safe.
Conservatively, we categorize a code fix as unsafe.
Unfortunately, this reduces the usefulness of the code fixes and encourages the use of unsafe fixes. 
In this regard, we should provide some guidelines and a clear procedure to update an eligible unsafe fix to a safe fix.

### Fix long-standing issues
  
- [x] [noUnreachableSuper] bugs ([rome#4674](https://github.com/rome/tools/issues/4674) and [4616](https://github.com/rome/tools/issues/4616))

- [x] [noUnnecessaryContinue] [rome#4745](https://github.com/rome/tools/issues/4745)

- [x] [noRedeclare] [rome#4291](https://github.com/rome/tools/issues/4291)

- [x] [noUnusedVariables] [rome#4245](https://github.com/rome/tools/issues/4245)

- [ ] [useSimplifiedLogicExpression] [rome#3527](https://github.com/rome/tools/issues/3527)

- [x] [useTemplate] incorrectly handle single-line comments [rome#4109](https://github.com/rome/tools/issues/4109)

- [ ] rome-ignore issue [rome#3734](https://github.com/rome/tools/issues/3734)

- [x] [useExponentiationOperator] [rome#3850](https://github.com/rome/tools/issues/3850) 

- [ ] Suppress comments and indentation [rome#3923](https://github.com/rome/tools/issues/3923)

### Test rules on big projects

- TypeScript Compiler
- [Ant Design](https://github.com/ant-design/ant-design)
- [surcase](https://github.com/alangpierce/sucrase)
- vscode code base
- deno std
- Any idea?

[noAccessKey]: https://biomejs.dev/linter/rules/no-access-key/
[noAccumulatingSpread]: https://biomejs.dev/linter/rules/no-accumulating-spread/
[noArguments]: https://biomejs.dev/linter/rules/no-arguments/
[noAriaUnsupportedElements]: https://biomejs.dev/linter/rules/no-aria-unsupported-elements/
[noArrayIndexKey]: https://biomejs.dev/linter/rules/no-array-index-key/
[noAssignInExpressions]: https://biomejs.dev/linter/rules/no-assign-in-expressions/
[noAsyncPromiseExecutor]: https://biomejs.dev/linter/rules/no-async-promise-executor/
[noAutofocus]: https://biomejs.dev/linter/rules/no-autofocus/
[noBannedTypes]: https://biomejs.dev/linter/rules/no-banned-types/
[noBlankTarget]: https://biomejs.dev/linter/rules/no-blank-target/
[noCatchAssign]: https://biomejs.dev/linter/rules/no-catch-assign/
[noChildrenProp]: https://biomejs.dev/linter/rules/no-children-prop/
[noClassAssign]: https://biomejs.dev/linter/rules/no-class-assign/
[noCommaOperator]: https://biomejs.dev/linter/rules/no-comma-operator/
[noCommentText]: https://biomejs.dev/linter/rules/no-comment-text/
[noCompareNegZero]: https://biomejs.dev/linter/rules/no-compare-neg-zero/
[noConfusingArrow]: https://biomejs.dev/linter/rules/no-confusing-arrow/
[noConfusingLabels]: https://biomejs.dev/linter/rules/no-confusing-labels/
[noConsoleLog]: https://biomejs.dev/linter/rules/no-console-log/
[noConstantCondition]: https://biomejs.dev/linter/rules/no-constant-condition/
[noConstAssign]: https://biomejs.dev/linter/rules/no-const-assign/
[noConstEnum]: https://biomejs.dev/linter/rules/no-const-enum/
[noConstructorReturn]: https://biomejs.dev/linter/rules/no-constructor-return/
[noControlCharactersInRegex]: https://biomejs.dev/linter/rules/no-control-characters-in-regex/
[noDangerouslySetInnerHtml]: https://biomejs.dev/linter/rules/no-dangerously-set-inner-html/
[noDangerouslySetInnerHtmlWithChildren]: https://biomejs.dev/linter/rules/no-dangerously-set-inner-html-with-children/
[noDebugger]: https://biomejs.dev/linter/rules/no-debugger/
[noDelete]: https://biomejs.dev/linter/rules/no-delete/
[noDistractingElements]: https://biomejs.dev/linter/rules/no-distracting-elements/
[noDoubleEquals]: https://biomejs.dev/linter/rules/no-double-equals/
[noDuplicateCase]: https://biomejs.dev/linter/rules/no-duplicate-case/
[noDuplicateClassMembers]: https://biomejs.dev/linter/rules/no-duplicate-class-members/
[noDuplicateJsonKeys]: https://biomejs.dev/linter/rules/no-duplicate-json-keys/
[noDuplicateJsxProps]: https://biomejs.dev/linter/rules/no-duplicate-jsx-props/
[noDuplicateObjectKeys]: https://biomejs.dev/linter/rules/no-duplicate-object-keys/
[noDuplicateParameters]: https://biomejs.dev/linter/rules/no-duplicate-parameters/
[noEmptyInterface]: https://biomejs.dev/linter/rules/no-empty-interface/
[noEmptyPattern]: https://biomejs.dev/linter/rules/no-empty-pattern/
[noExcessiveComplexity]: https://biomejs.dev/linter/rules/no-excessive-complexity/
[noExplicitAny]: https://biomejs.dev/linter/rules/no-explicit-any/
[noExtraBooleanCast]: https://biomejs.dev/linter/rules/no-extra-boolean-cast/
[noExtraNonNullAssertion]: https://biomejs.dev/linter/rules/no-extra-non-null-assertion/
[noFallthroughSwitchClause]: https://biomejs.dev/linter/rules/no-fallthrough-switch-clause/
[noForEach]: https://biomejs.dev/linter/rules/no-for-each/
[noFunctionAssign]: https://biomejs.dev/linter/rules/no-function-assign/
[noGlobalIsFinite]: https://biomejs.dev/linter/rules/no-global-is-finite/
[noGlobalIsNan]: https://biomejs.dev/linter/rules/no-global-is-nan/
[noGlobalObjectCalls]: https://biomejs.dev/linter/rules/no-global-object-calls/
[noHeaderScope]: https://biomejs.dev/linter/rules/no-header-scope/
[noImplicitBoolean]: https://biomejs.dev/linter/rules/no-implicit-boolean/
[noImportAssign]: https://biomejs.dev/linter/rules/no-import-assign/
[noInferrableTypes]: https://biomejs.dev/linter/rules/no-inferrable-types/
[noInnerDeclarations]: https://biomejs.dev/linter/rules/no-inner-declarations/
[noInvalidConstructorSuper]: https://biomejs.dev/linter/rules/no-invalid-constructor-super/
[noLabelVar]: https://biomejs.dev/linter/rules/no-label-var/
[noMultipleSpacesInRegularExpressionLiterals]: https://biomejs.dev/linter/rules/no-multiple-spaces-in-regular-expression-literals/
[noNamespace]: https://biomejs.dev/linter/rules/no-namespace/
[noNegationElse]: https://biomejs.dev/linter/rules/no-negation-else/
[noNewSymbol]: https://biomejs.dev/linter/rules/no-new-symbol/
[noNoninteractiveElementToInteractiveRole]: https://biomejs.dev/linter/rules/no-noninteractive-element-to-interactive-role/
[noNoninteractiveTabindex]: https://biomejs.dev/linter/rules/no-noninteractive-tabindex/
[noNonNullAssertion]: https://biomejs.dev/linter/rules/no-non-null-assertion/
[noNonoctalDecimalEscape]: https://biomejs.dev/linter/rules/no-nonoctal-decimal-escape/
[noParameterAssign]: https://biomejs.dev/linter/rules/no-parameter-assign/
[noParameterProperties]: https://biomejs.dev/linter/rules/no-parameter-properties/
[noPositiveTabindex]: https://biomejs.dev/linter/rules/no-positive-tabindex/
[noPrecisionLoss]: https://biomejs.dev/linter/rules/no-precision-loss/
[noPrototypeBuiltins]: https://biomejs.dev/linter/rules/no-prototype-builtins/
[noRedeclare]: https://biomejs.dev/linter/rules/no-redeclare/
[noRedundantAlt]: https://biomejs.dev/linter/rules/no-redundant-alt/
[noRedundantRoles]: https://biomejs.dev/linter/rules/no-redundant-roles/
[noRedundantUseStrict]: https://biomejs.dev/linter/rules/no-redundant-use-strict/
[noRenderReturnValue]: https://biomejs.dev/linter/rules/no-render-return-value/
[noRestrictedGlobals]: https://biomejs.dev/linter/rules/no-restricted-globals/
[noSelfAssign]: https://biomejs.dev/linter/rules/no-self-assign/
[noSelfCompare]: https://biomejs.dev/linter/rules/no-self-compare/
[noSetterReturn]: https://biomejs.dev/linter/rules/no-setter-return/
[noShadowRestrictedNames]: https://biomejs.dev/linter/rules/no-shadow-restricted-names/
[noShoutyConstants]: https://biomejs.dev/linter/rules/no-shouty-constants/
[noSparseArray]: https://biomejs.dev/linter/rules/no-sparse-array/
[noStaticOnlyClass]: https://biomejs.dev/linter/rules/no-static-only-class/
[noStringCaseMismatch]: https://biomejs.dev/linter/rules/no-string-case-mismatch/
[noSvgWithoutTitle]: https://biomejs.dev/linter/rules/no-svg-without-title/
[noSwitchDeclarations]: https://biomejs.dev/linter/rules/no-switch-declarations/
[noUndeclaredVariables]: https://biomejs.dev/linter/rules/no-undeclared-variables/
[noUnnecessaryContinue]: https://biomejs.dev/linter/rules/no-unnecessary-continue/
[noUnreachable]: https://biomejs.dev/linter/rules/no-unreachable/
[noUnreachableSuper]: https://biomejs.dev/linter/rules/no-unreachable-super/
[noUnsafeDeclarationMerging]: https://biomejs.dev/linter/rules/no-unsafe-declaration-merging/
[noUnsafeFinally]: https://biomejs.dev/linter/rules/no-unsafe-finally/
[noUnsafeNegation]: https://biomejs.dev/linter/rules/no-unsafe-negation/
[noUnsafeOptionalChaining]: https://biomejs.dev/linter/rules/no-unsafe-optional-chaining/
[noUnusedLabels]: https://biomejs.dev/linter/rules/no-unused-labels/
[noUnusedTemplateLiteral]: https://biomejs.dev/linter/rules/no-unused-template-literal/
[noUnusedVariables]: https://biomejs.dev/linter/rules/no-unused-variables/
[noUselessCatch]: https://biomejs.dev/linter/rules/no-useless-catch/
[noUselessConstructor]: https://biomejs.dev/linter/rules/no-useless-constructor/
[noUselessEmptyExport]: https://biomejs.dev/linter/rules/no-useless-empty-export/
[noUselessFragments]: https://biomejs.dev/linter/rules/no-useless-fragments/
[noUselessLabel]: https://biomejs.dev/linter/rules/no-useless-label/
[noUselessRename]: https://biomejs.dev/linter/rules/no-useless-rename/
[noUselessSwitchCase]: https://biomejs.dev/linter/rules/no-useless-switch-case/
[noUselessThisAlias]: https://biomejs.dev/linter/rules/no-useless-this-alias/
[noUselessTypeConstraint]: https://biomejs.dev/linter/rules/no-useless-type-constraint/
[noVar]: https://biomejs.dev/linter/rules/no-var/
[noVoid]: https://biomejs.dev/linter/rules/no-void/
[noVoidElementsWithChildren]: https://biomejs.dev/linter/rules/no-void-elements-with-children/
[noVoidTypeReturn]: https://biomejs.dev/linter/rules/no-void-type-return/
[noWith]: https://biomejs.dev/linter/rules/no-with/
[useAltText]: https://biomejs.dev/linter/rules/use-alt-text/
[useAnchorContent]: https://biomejs.dev/linter/rules/use-anchor-content/
[useAriaPropsForRole]: https://biomejs.dev/linter/rules/use-aria-props-for-role/
[useAriaPropTypes]: https://biomejs.dev/linter/rules/use-aria-prop-types/
[useArrowFunction]: https://biomejs.dev/linter/rules/use-arrow-function/
[useBlockStatements]: https://biomejs.dev/linter/rules/use-block-statements/
[useButtonType]: https://biomejs.dev/linter/rules/use-button-type/
[useConst]: https://biomejs.dev/linter/rules/use-const/
[useDefaultParameterLast]: https://biomejs.dev/linter/rules/use-default-parameter-last/
[useDefaultSwitchClauseLast]: https://biomejs.dev/linter/rules/use-default-switch-clause-last/
[useEnumInitializers]: https://biomejs.dev/linter/rules/use-enum-initializers/
[useExhaustiveDependencies]: https://biomejs.dev/linter/rules/use-exhaustive-dependencies/
[useExponentiationOperator]: https://biomejs.dev/linter/rules/use-exponentiation-operator/
[useFlatMap]: https://biomejs.dev/linter/rules/use-flat-map/
[useFragmentSyntax]: https://biomejs.dev/linter/rules/use-fragment-syntax/
[useGetterReturn]: https://biomejs.dev/linter/rules/use-getter-return/
[useGroupedTypeImport]: https://biomejs.dev/linter/rules/use-grouped-type-import/
[useHeadingContent]: https://biomejs.dev/linter/rules/use-heading-content/
[useHookAtTopLevel]: https://biomejs.dev/linter/rules/use-hook-at-top-level/
[useHtmlLang]: https://biomejs.dev/linter/rules/use-html-lang/
[useIframeTitle]: https://biomejs.dev/linter/rules/use-iframe-title/
[useImportRestrictions]: https://biomejs.dev/linter/rules/use-import-restrictions/
[useIsArray]: https://biomejs.dev/linter/rules/use-is-array/
[useIsNan]: https://biomejs.dev/linter/rules/use-is-nan/
[useKeyWithClickEvents]: https://biomejs.dev/linter/rules/use-key-with-click-events/
[useKeyWithMouseEvents]: https://biomejs.dev/linter/rules/use-key-with-mouse-events/
[useLiteralEnumMembers]: https://biomejs.dev/linter/rules/use-literal-enum-members/
[useLiteralKeys]: https://biomejs.dev/linter/rules/use-literal-keys/
[useMediaCaption]: https://biomejs.dev/linter/rules/use-media-caption/
[useNamespaceKeyword]: https://biomejs.dev/linter/rules/use-namespace-keyword/
[useNamingConvention]: https://biomejs.dev/linter/rules/use-naming-convention/
[useNumericLiterals]: https://biomejs.dev/linter/rules/use-numeric-literals/
[useOptionalChain]: https://biomejs.dev/linter/rules/use-optional-chain/
[useSelfClosingElements]: https://biomejs.dev/linter/rules/use-self-closing-elements/
[useShorthandArrayType]: https://biomejs.dev/linter/rules/use-shorthand-array-type/
[useSimpleNumberKeys]: https://biomejs.dev/linter/rules/use-simple-number-keys/
[useSimplifiedLogicExpression]: https://biomejs.dev/linter/rules/use-simplified-logic-expression/
[useSingleCaseStatement]: https://biomejs.dev/linter/rules/use-single-case-statement/
[useSingleVarDeclarator]: https://biomejs.dev/linter/rules/use-single-var-declarator/
[useTemplate]: https://biomejs.dev/linter/rules/use-template/
[useValidAnchor]: https://biomejs.dev/linter/rules/use-valid-anchor/
[useValidAriaProps]: https://biomejs.dev/linter/rules/use-valid-aria-props/
[useValidForDirection]: https://biomejs.dev/linter/rules/use-valid-for-direction/
[useValidLang]: https://biomejs.dev/linter/rules/use-valid-lang/
[useValidTypeof]: https://biomejs.dev/linter/rules/use-valid-typeof/
[useWhile]: https://biomejs.dev/linter/rules/use-while/
[useYield]: https://biomejs.dev/linter/rules/use-yield/

**Comment by JoshuaKGoldberg** [comment]
> Test rules on big projects

Suggestion: test the rules on other AST / static analysis projects, and linting projects (ESLint, typescript-eslint, etc) in particular.

* AST projects tend to have a lot of structures / pieces of syntax that are different from your typical userland ones
* Looking at linters raises visibility for what everyone else is doing, which can be good for feature & edge case ideation

**Comment by smorimoto** [comment]
My team explicitly enabled this, but I'm still not sure how hard this is to port.
https://typescript-eslint.io/rules/switch-exhaustiveness-check

**Reply by ematipico** [reply]
It requires type checking, which Biome doesn't have yet unfortunately

**Reply by smorimoto** [reply]
That's where I was concerned.

**Reply by jpike88** [reply]
Desperate for this one... would this be a low hanging fruit for when type awareness is brought? I guess it's just checking that all items in the union of the type have a matching case right

**Reply by arendjr** [reply]
I wouldn’t say it’s exactly trivial, because you may have a switch over something like`object.type`, where `object` has union type and `type` has a derived union type. So we’ll need to be at a point where we have derived types at least. But once we have those, this rule shouldn’t be too hard anymore.

Btw, a trick I tend to use today is to make sure switches are always in a standalone function with an explicit return type annotation. This way, I can use `return` instead of `break` and TypeScript will complain if I forget a `case`. Although this assumes `undefined` is not an acceptable return value, because this strategy doesn’t help if it is.

**Comment by smorimoto** [comment]
This is very important when implementing software while using Node.js compatibility mode in the [Edge Runtime](https://edge-runtime.vercel.app), and if it's supported by default, it's much better.
https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v49.0.0/docs/rules/prefer-node-protocol.md

**Reply by Conaclos** [reply]
[useNodeImportProtocol](https://biomejs.dev/linter/rules/use-nodejs-import-protocol/) is implemented and will be available in the upcoming release v1.5.0

**Reply by smorimoto** [reply]
Superb!!

**Reply by smorimoto** [reply]
@Conaclos By the way, does it only support ESM? Or does it also support CJS?

**Reply by Conaclos** [reply]
I also supports `require` imports. You can [try it on the playground](https://biomejs.dev/playground/?lintRules=all&code=YwBvAG4AcwB0ACAAeAAgAD0AIAByAGUAcQB1AGkAcgBlACgAIgBmAHMAIgApADsA).

**Comment by smorimoto** [comment]
I believe these are good directions for performance and consistency.
https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v49.0.0/docs/rules/prefer-set-has.md
https://github.com/sindresorhus/eslint-plugin-unicorn/blob/v49.0.0/docs/rules/prefer-spread.md

**Reply by Conaclos** [reply]
Please use the [`✅ Linter rules from other sources`](https://github.com/biomejs/biome/discussions/3) disussion to propose rules.

**Comment by vasucp1207** [comment]
I have a question, is it a good idea for rules like `useAltText`, `useButtonType` to provide a code fix with an empty string to the missing attribute?

Ex for `useAltText`,
```js
<img src="image.png" />
```

After applying code fix
```js
<img src="image.png" alt="" />
```

**Reply by Conaclos** [reply]
I am not sure if it is a valuable fix? Maybe for IDE? We could provide this as an unsafe fix. What do you think @ematipico ?

**Reply by unvalley** [reply]
@vasucp1207 
I had a similar thought when I implemented `noSvgWithoutTitle`. If incomplete suggestions are accepted, users who accept without much thought may make the mistake of judging that the problem has been solved. It would be nice if the empty-string state would also produce a lint in the case, because it's still incomplete.

**Reply by vasucp1207** [reply]
Yeah, checking for empty strings can be very helpful, we should provide a different diagnostic message in that case.

**Reply by Conaclos** [reply]
It seems confusing to me that a rule applies a fix and then complains about the fix it applied... In this case it is better not to provide any fix. Still, it is a good idea to print a diagnostic for empty strings.

**Reply by smorimoto** [reply]
I think we should pay attention to these, but the direction is right.

- https://www.w3.org/TR/2016/NOTE-WCAG20-TECHS-20161007/H37
- https://www.w3.org/TR/html401/struct/objects.html#adef-alt
  - Do not specify irrelevant alternate text when including images intended to format a page, for instance, `alt="red ball"` would be inappropriate for an image that adds a red ball for decorating a heading or paragraph. In such cases, the alternate text should be the empty string (`""`). Authors are in any case advised to avoid using images to format pages; style sheets should be used instead.
  - Do not specify meaningless alternate text (e.g., "dummy text"). Not only will this frustrate users, it will slow down user agents that must convert text to speech or braille output.

**Reply by vasucp1207** [reply]
Right providing a fix doesn't make sense if it's not going to solve the problem completely. But the diagnostic can help.
